diff --git a/config/BUILDCONFIG.gn b/config/BUILDCONFIG.gn
index e011502..ae5758e 100644
--- a/config/BUILDCONFIG.gn
+++ b/config/BUILDCONFIG.gn
@@ -453,10 +453,10 @@ if (is_win) {
 }
 
 if (is_posix) {
-  if (current_os != "aix") {
-    default_compiler_configs +=
-        [ "//build/config/gcc:symbol_visibility_hidden" ]
-  }
+  #if (current_os != "aix") {
+  #  default_compiler_configs +=
+  #      [ "//build/config/gcc:symbol_visibility_hidden" ]
+  #}
 }
 
 if (is_fuchsia) {
